src/StreamingBodyParser/FormUrlencodedParser.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace React\Http\StreamingBodyParser;
+
+use Evenement\EventEmitter;
+use Psr\Http\Message\RequestInterface;
+use React\Http\HttpBodyStream;
+
+final class FormUrlencodedParser extends EventEmitter
+{
+    /**
+     * @var RequestInterface
+     */
+    private $request;
+
+    /**
+     * @var HttpBodyStream
+     *
+     * @internal
+     */
+    public $body;
+
+    /**
+     * @var string
+     *
+     * @internal
+     */
+    public $buffer = '';
+
+    /**
+     * @param RequestInterface $request
+     */
+    public function __construct(RequestInterface $request)
+    {
+        $this->request = $request;
+        $this->body = $this->request->getBody();
+        $this->body->on('data', array($this, 'onData'));
+        $that = $this;
+        $this->body->on('end', function () use ($that) {
+            $that->body->removeAllListeners();
+            $that->parse($that->buffer);
+            $that->emit('end');
+        });
+    }
+
+    /**
+     * @internal
+     */
+    public function onData($data)
+    {
+        $this->buffer .= $data;
+
+        $pos = strrpos($this->buffer, '&');
+        if ($pos === false) {
+            return;
+        }
+
+        $buffer = substr($this->buffer, 0, $pos);
+        $this->buffer = substr($this->buffer, $pos + 1);
+
+        $this->parse($buffer);
+    }
+
+    /**
+     * @internal
+     */
+    public function parse($buffer)
+    {
+        foreach (explode('&', $buffer) as $chunk) {
+            $this->emit(
+                'post',
+                explode('=', rawurldecode($chunk))
```

Comment:
Not sure, but maybe ensure here that there will be always an array with to entries (kay and value) emitted. so that `key1&key2` emits `array('key1', null)` and `array('key2', null`)`?